plugins/synaptics-vmm9/fu-synaptics-vmm9-device.c

```diff
@@ -0,0 +1,709 @@
+	if (!fu_input_stream_chunkify(stream,
+				      fu_synaptics_vmm9_device_crc16_cb,
+				      &crc16_image,
+				      error)) {
+		g_prefix_error(error, "failed to compute image CRC16: ");
+		return FALSE;
+	}
+
+	/* erase the storage bank */
+	if (!fu_synaptics_vmm9_device_erase(self, fu_progress_get_child(progress), error)) {
+		g_prefix_error(error, "failed to erase: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* ensure the SPI flash is ready to access the write command */
+	fu_device_sleep_full(device, 3000, fu_progress_get_child(progress));
+	fu_progress_step_done(progress);
+
+	/* write each block */
+	chunks =
+	    fu_chunk_array_new_from_stream(stream, 0x0, FU_STRUCT_HID_PAYLOAD_SIZE_FIFO, error);
+	if (chunks == NULL)
+		return FALSE;
+	if (!fu_synaptics_vmm9_device_write_blocks(self,
+						   chunks,
+						   fu_progress_get_child(progress),
+						   error)) {
+		g_prefix_error(error, "failed to write: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* verify each block */
+	if (!fu_synaptics_vmm9_device_verify_crc16(self, crc16_image, streamsz, error))
+		return FALSE;
+	fu_progress_step_done(progress);
+
+	/* activate the firmware */
+	if (!fu_synaptics_vmm9_device_command(self,
+					      FU_SYNAPTICS_VMM9_RC_CTRL_ACTIVATE_FIRMWARE,
+					      0x0,
+					      NULL,
+					      0,
+					      NULL,
+					      0,
+					      FU_SYNAPTICS_VMM9_COMMAND_FLAG_NONE,
+					      error)) {
+		g_prefix_error(error, "failed to activate: ");
+		return FALSE;
+	}
+	fu_progress_step_done(progress);
+
+	/* success! */
+	return TRUE;
+}
+
+static void
+fu_synaptics_vmm9_device_set_progress(FuDevice *self, FuProgress *progress)
+{
+	fu_progress_set_id(progress, G_STRLOC);
+	fu_progress_add_flag(progress, FU_PROGRESS_FLAG_GUESSED);
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "detach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_WRITE, 57, "write");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_RESTART, 0, "attach");
+	fu_progress_add_step(progress, FWUPD_STATUS_DEVICE_BUSY, 43, "reload");
+}
+
+static void
+fu_synaptics_vmm9_device_init(FuSynapticsVmm9Device *self)
+{
+	// fu_device_set_firmware_size(FU_DEVICE(self), 64 * 1024 * 5);
+	fu_device_set_firmware_size(FU_DEVICE(self), 0x80000);
+	fu_device_set_version_format(FU_DEVICE(self), FWUPD_VERSION_FORMAT_TRIPLET);
+	fu_device_set_remove_delay(FU_DEVICE(self), FU_DEVICE_REMOVE_DELAY_RE_ENUMERATE);
+	fu_device_add_protocol(FU_DEVICE(self), "com.synaptics.mst");
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_CAN_VERIFY_IMAGE);
+	fu_device_add_flag(FU_DEVICE(self), FWUPD_DEVICE_FLAG_SIGNED_PAYLOAD);
+	fu_device_add_internal_flag(FU_DEVICE(self), FU_DEVICE_INTERNAL_FLAG_ONLY_WAIT_FOR_REPLUG);
+	fu_device_register_private_flag(FU_DEVICE(self),
+					FU_SYNAPTICS_VMM9_DEVICE_FLAG_MANUAL_RESTART_REQUIRED,
+					"manual-restart-required");
```

Comment:
this is pretty annoying, is there any chance to add a command to restart it?  So we can at least have an activation flow.

Reply:
Yup, I've asked Synaptics about this.

Reply:
Yup, I traced out VmmHidTool and there's one register write to reset the board. Fixed.